### DISCUSSION TITLE
LSP find reference ui is super bad

### DISCUSSION THREAD

**Discussion by 0x8f701** [neovim/neovim#30604] [General]
### Problem

no preview window, no highlight, doesn't automatically close the quicklist after hit enter

### Expected behavior

1. esc return back to original position and close the quickfix window
2. searched symbol should be highlighted(otherwise user still need to find where is the referenced item)
3. should have a simpler mapping like using j/k to move around in quickfix window (main window should automatically jumps to that file)

**Comment by 0x8f701** [comment]
The current Neovim LSP experience is quite poor compared to coc.nvim. The configuration is scattered across different plugins and different places, and you have to piece them together manually. While some might consider this is "decentralized", I believe the basic, core LSP functionalities should be centralized in one place, or even built-in. Or at least a recommend config should be given out to contain all the features. Now there is just no such recommanded config. People just spend tons of time reading blogs, ask chatgpts, readiing plugin readmes.

LSP is a complex and highly coupled  thing, it will consume the users too much time improve the default behaviour.

**Comment by justinmk** [comment]
> doesn't automatically close the quicklist after hit enter

Why do you assume everyone wants this?

The rest of this issue is based on unfounded assumptions from the perspective that the builtin quickfix interface isn't good. It's actually very usable when paired with https://github.com/tpope/vim-unimpaired (which has some mappings we should include in Nvim by default).

Starting with the assumption that the UI needs to match coc.nvim or vscode is a fallacy. There are many parts of vim such as quickfix which have tons of features, but require some investment. That is the general theme of vim in general, and if you don't think it's useful then there is no point in using vim. (Note that this is 100% orthogonal to monotonic improvements, i.e. improvements that can be made to Nvim defaults or features without throwing away the useful parts)

**Comment by 0x8f701** [comment]
@justinmk please be aware that there are other issues even if you pointed that maybe the others don't want to automatically close:

> no preview window, no highlight

I don't think this issue should be closed. I didn't assume it should match with vscode or coc.nvim but it's just tooooo bad experience.

Even if you ask GPT, it will still tell you that the user experience of coc.nvim is better than the current Neovim LSP, and it's more ready-to-use out of the box. Why don't learn from them on this part?

**Comment by justinmk** [comment]
> no preview window, no highlight

what is meant by "no highlight"? there is various kinds of highlighting. If you meant syntax highlighting in the quickfix , that isn't specific to LSP. There are existing issues tracking improvements to quickfix: https://github.com/neovim/neovim/issues/15950

OTOH, syntax highlight in quickfix could be a performance cost, so that would need to be weighed. Meanwhile, it's easy to install plugins that do this...

> no preview window

we have existing issues tracking enhanced "preview" functionality: https://github.com/neovim/neovim/issues/9920

that kind of generalized functionality is required, we aren't going to make non-trivial ad-hoc changes to LSP "find references".

> Even if you ask GPT, it will still tell you that the user experience of coc.nvim is better than the current Neovim LSP

GPT gets its opinions from the internet. wtf is that supposed to prove?

**Comment by 0x8f701** [comment]
> what is meant by "no highlight"? 

no highlight/cursorline on the main window, syntax highlight is not important, but we should be able to match quickfix's current line with the main window. Now it's difficult to see which line it maps to

it's also ok to not have preview window, as long as the mapping works then it should be ok.


> why am I saying it's too bad?
because if you cnext in quicklist window, because there is no preview window, no mapping, it has to jump to that pos of that file...people easily get lost. ESC cannot make sure the user is on the original file anymore. This is extremely bad!

**Reply by justinmk** [reply]
> it has to jump to that pos of that file...people easily get lost. ESC cannot make sure the user is on the original file anymore.

There are all kinds of navigation features that address this. ctrl-t, ctrl-o, uppercase marks.

**Reply by justinmk** [reply]
> we should be able to match quickfix's current line with the main window. Now it's difficult to see which line it maps to

`QuickFixLine` is highlighted, but I agree it's too subtle currently.

**Reply by 0x8f701** [reply]
will learn and use quickfix better, thanks~

**Comment by 0x8f701** [comment]
@justinmk 

find reference triggered on "vim", please see the diff, in coc I can easily hit Esc to quit without jumping to any position. in lua I got lost because 


first, I can forget what word am I searching on

second it directly jumps to that file (which is super super bad because if I don't want then I need to manually jump to original file, sometimes can be hard)

third no preview window (which caused previous issue, it has to jump to the file to display it)

my coc config:
https://github.com/GopherJ/dotfiles/tree/master/coc/v1
![image](https://github.com/user-attachments/assets/bbbf9f36-7ab8-4788-8703-c18eddabf134)


my lua config(do you know what am I searching on? why it jumps to the file instead of displaying it in prewview!!!) :
https://github.com/GopherJ/dotfiles/tree/master/lua/v0
![image](https://github.com/user-attachments/assets/d4475b44-4596-4749-9346-94df6bbad98d)

current find reference config:
https://github.com/GopherJ/dotfiles/blob/1339c39fdbf854ed6741987d1ad389a7401b059d/lua/v0/lua/vimlua/plugins/completion/init.lua#L33-L64

It's also ok to not have preview window as long as we can esily go back to previous position and there is highlight on the search keyword

**Reply by justinmk** [reply]
> I can easily hit Esc to quit without jumping to any position.

That just means you didn't learn about nvim's navigation features. I don't want a popup for quickfix results, because I know how to navigate and that gives more flexibility.

**Reply by justinmk** [reply]
> second it directly jumps to that file (which is super super bad because if I don't want then I need to manually jump to original file, sometimes can be hard)

Use `ctrl-w<cr>` in the quickfix window to open the result in a split. Then you can close the split.

**Reply by 0x8f701** [reply]
Thank you for telling me about this. it's very true I don't often use quickfix or tabs. I just tried this method, and while it solves some problems, it is still very inefficient.

**Comment by 0x8f701** [comment]
Another thing is that I estimate people will need to spend several times more time configuring neovim lsp in order to achieve 60% of the functionality described in the coc.nvim README.

**Reply by justinmk** [reply]
> people will need to spend several times more time configuring neovim lsp in order to achieve 60% of the functionality described in the coc.nvim README.

Those people haven't bothered to learn nvim's existing features and are looking for a more vscode-like experience.

**Reply by 0x8f701** [reply]
Not really, I rarely use VSCode, but I use coc.nvim and have been using it for four years, I love its way of finding references

**Reply by clason** [reply]
You are very welcome to keep using that, you know.

**Reply by clason** [reply]
> Expected behavior

> the same as coc.nvim

It is a valid response to that. Obviously there are concrete improvements we could make, but you can't complain about Neovim not being something it explicitly is not meant to be. (After all, coc.nvim already exists.)

**Comment by 0x8f701** [comment]
BTW I use this now, close to what I'm used to 

```lua

local function safe_cmd(cmd)
	local success, _ = pcall(vim.cmd, cmd)
	if not success then
		return false
	end
	return true
end

vim.api.nvim_create_autocmd("FileType", {
pattern = "qf",
callback = function()
	local qf_list = vim.fn.getqflist()
	local qf_length = #qf_list
	local qf_winid = vim.fn.win_getid(vim.fn.winnr("$"))
	local main_winid = vim.fn.win_getid(vim.fn.winnr("#"))
	local main_bufnr = vim.fn.bufnr("#")
	local main_cursor = vim.api.nvim_win_get_cursor(main_winid)

	vim.keymap.set("n", "j", function()
		local current_idx = vim.fn.line(".")
		if current_idx < qf_length then
			vim.schedule(function()
				if safe_cmd("cnext") and vim.api.nvim_win_is_valid(qf_winid) then
					vim.api.nvim_set_current_win(qf_winid)
				end
			end)
		end
	end, { buffer = true, silent = true })

	vim.keymap.set("n", "k", function()
		local current_idx = vim.fn.line(".")
		if current_idx > 1 then
			vim.schedule(function()
				if safe_cmd("cprev") and vim.api.nvim_win_is_valid(qf_winid) then
					vim.api.nvim_set_current_win(qf_winid)
				end
			end)
		end
	end, { buffer = true, silent = true })

	vim.api.nvim_buf_set_keymap(
		0,
		"n",
		"<CR>",
		"<CR><cmd>cclose<CR>",
		{ noremap = true, silent = true }
	)

	vim.keymap.set("n", "<ESC>", function()
		vim.cmd("cclose")
		vim.api.nvim_set_current_buf(main_bufnr)
		vim.api.nvim_win_set_cursor(main_winid, main_cursor)
	end, { buffer = true, silent = true })
end,
})

```

**Reply by justinmk** [reply]
note that your "j"/"k" maps are not needed since Nvim now has `]q` and `[q` by default (and they work *globally*, not only in the quickfix window): https://github.com/neovim/neovim/pull/28525

**Comment by 0x8f701** [comment]
**EDIT**: added highlight

```lua
-- init.vim
highlight QuickFixHL ctermfg=white ctermbg=red cterm=bold guifg=white guibg=red gui=bold

-- utils.lua
function M.clear_highlight(winid, var_id)
	local success, highlight_id = pcall(vim.api.nvim_win_get_var, winid, var_id)
	if success then
		pcall(vim.fn.matchdelete, highlight_id, winid)
		vim.api.nvim_win_del_var(winid, var_id)
	end
end

function M.highlight_item(hl_group, winid, item, var_id)
	M.clear_highlight(winid, var_id)

	local lnum = item.lnum
	local start_col = math.max(1, item.col or 1)
	local end_col = item.end_col
	local pattern

	if item.user_data and item.user_data.range and item.user_data.range["end"] then
		local user_end_col = item.user_data.range["end"].character + 1
		if user_end_col > start_col then
			end_col = user_end_col
		end
	end

	if not end_col or end_col <= start_col then
		pattern = string.format("\\%%%dl", lnum)
	else
		pattern = string.format("\\%%%dl\\%%%dc.*\\%%%dc", lnum, start_col, end_col)
	end

	local match_id = vim.fn.matchadd(hl_group, pattern, 10, -1, { window = winid })
	vim.api.nvim_win_set_var(winid, var_id, match_id)
end

return M


-- your lsp lua file
vim.api.nvim_create_autocmd("FileType", {
pattern = "qf",
callback = function()
	local qf_list = vim.fn.getqflist()
	local qf_length = #qf_list
	local qf_winid = vim.api.nvim_get_current_win()
	local main_winid = vim.fn.win_getid(vim.fn.winnr("#"))
	local main_bufnr = vim.fn.bufnr("#")
	local main_cursor = vim.api.nvim_win_get_cursor(main_winid)

	if #qf_list > 0 then
		utils.highlight_item("QuickFixHL", main_winid, qf_list[1], "quickfix_highlight_id")
	end

	vim.keymap.set("n", "j", function()
		local current_idx = vim.fn.line(".")
		if current_idx < qf_length then
			local item = qf_list[current_idx + 1]
			utils.highlight_item("QuickFixHL", main_winid, item, "quickfix_highlight_id")

			vim.schedule(function()
				if utils.safe_cmd("cnext") and vim.api.nvim_win_is_valid(qf_winid) then
					vim.api.nvim_set_current_win(qf_winid)
				end
			end)
		end
	end, { buffer = true, silent = true })

	vim.keymap.set("n", "k", function()
		local current_idx = vim.fn.line(".")
		if current_idx > 1 then
			local item = qf_list[current_idx - 1]
			utils.highlight_item("QuickFixHL", main_winid, item, "quickfix_highlight_id")

			vim.schedule(function()
				if utils.safe_cmd("cprev") and vim.api.nvim_win_is_valid(qf_winid) then
					vim.api.nvim_set_current_win(qf_winid)
				end
			end)
		end
	end, { buffer = true, silent = true })

	vim.keymap.set("n", "<CR>", function()
		vim.fn.feedkeys(vim.api.nvim_replace_termcodes("<CR>", true, false, true), "n")
		vim.defer_fn(function()
			vim.cmd("cclose")

			utils.clear_highlight(main_winid, "quickfix_highlight_id")
		end, 10)
	end, { noremap = true, silent = true })

	vim.keymap.set("n", "<ESC>", function()
		vim.cmd("cclose")
		if vim.api.nvim_buf_is_valid(main_bufnr) then
			vim.api.nvim_set_current_buf(main_bufnr)
		end
		if vim.api.nvim_win_is_valid(main_winid) then
			vim.api.nvim_win_set_cursor(main_winid, main_cursor)
			utils.clear_highlight(main_winid, "quickfix_highlight_id")
		end
	end, { buffer = true, silent = true })
end,
})
```
